Flatten the Tritium proxy to avoid excessive stack frames

carterkozak opened the pull request:
Note that this pushes the invoke method above the 35 byte inline
threshold to 102 bytes, which is still below the 325 byte hot
method threshold.

Before:
```
java.lang.RuntimeException: Stack Trace
        at com.palantir.DemoMain.run(DemoMain.java:32)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at com.palantir.tritium.proxy.InvocationEventProxy.execute(InvocationEventProxy.java:149)
        at com.palantir.tritium.proxy.InvocationEventProxy.invoke(InvocationEventProxy.java:140)
        at com.palantir.tritium.proxy.InvocationEventProxy.instrumentInvocation(InvocationEventProxy.java:120)
        at com.palantir.tritium.proxy.InvocationEventProxy.handleInvocation(InvocationEventProxy.java:99)
        at com.google.common.reflect.AbstractInvocationHandler.invoke(AbstractInvocationHandler.java:86)
        at com.sun.proxy.$Proxy0.run(Unknown Source)
        at com.palantir.DemoMain.main(DemoMain.java:27)
```

After:
```
java.lang.RuntimeException: Stack Trace
        at com.palantir.DemoMain.run(DemoMain.java:32)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at com.palantir.tritium.proxy.InvocationEventProxy.invoke(InvocationEventProxy.java:102)
        at com.sun.proxy.$Proxy0.run(Unknown Source)
        at com.palantir.DemoMain.main(DemoMain.java:27)
```

## After this PR
==COMMIT_MSG==
Flatten the Tritium proxy to avoid excessive stack frames
==COMMIT_MSG==

carterkozak committed:
Flatten the Tritium proxy to avoid excessive stack frames

Note that this pushes the invoke method above the 35 byte inline
threshold to 102 bytes, which is still below the 325 byte hot
method threshiold.

Before:
```
java.lang.RuntimeException: Stack Trace
        at com.palantir.DemoMain.run(DemoMain.java:32)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at com.palantir.tritium.proxy.InvocationEventProxy.execute(InvocationEventProxy.java:149)
        at com.palantir.tritium.proxy.InvocationEventProxy.invoke(InvocationEventProxy.java:140)
        at com.palantir.tritium.proxy.InvocationEventProxy.instrumentInvocation(InvocationEventProxy.java:120)
        at com.palantir.tritium.proxy.InvocationEventProxy.handleInvocation(InvocationEventProxy.java:99)
        at com.google.common.reflect.AbstractInvocationHandler.invoke(AbstractInvocationHandler.java:86)
        at com.sun.proxy.$Proxy0.run(Unknown Source)
        at com.palantir.DemoMain.main(DemoMain.java:27)
```

After:
```
java.lang.RuntimeException: Stack Trace
        at com.palantir.DemoMain.run(DemoMain.java:32)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at com.palantir.tritium.proxy.InvocationEventProxy.invoke(InvocationEventProxy.java:102)
        at com.sun.proxy.$Proxy0.run(Unknown Source)
        at com.palantir.DemoMain.main(DemoMain.java:27)
```

carterkozak committed:
Add generated changelog entries

schlosna commented on tritium-lib/src/main/java/com/palantir/tritium/proxy/InvocationEventProxy.java:
```diff
@@ -88,39 +88,62 @@ private boolean isEnabled(Object instance, Method method, Object[] args) {
+        }
+        if (isEquals(method, arguments)) {
+            Object arg = arguments[0];
+            return arg != null && proxy == arg;
```
This is changing equals semantics compared to [Guava AbstractInvocationHandler](https://github.com/google/guava/blob/master/guava/src/com/google/common/reflect/AbstractInvocationHandler.java#L70-L81) , which we may want to maintain.  

While evaluating `equals` on a proxy is odd, folks have made this mistake before, e.g. atlasdb at one point where they had a hash map of proxies, and switched to an identity hash map https://github.com/palantir/atlasdb/blob/1962cc6e650f2944f36ea88aa9b91df5133261f4/atlasdb-config/src/main/java/com/palantir/atlasdb/factory/Leaders.java#L226-L241

carterkozak replied:
I don't believe tritium allows us to create multiple proxy instances with the same invocation handler instance, nor do we implement a custom invocationhandler equals. I will add tests and make sure they pass before and after my change.

schlosna commented on tritium-lib/src/test/java/com/palantir/tritium/proxy/InvocationEventProxyTest.java:
```diff
@@ -300,8 +299,9 @@ private static InvocationEventProxy createTestProxy(InvocationEventHandler<Invoc
 
-    private static Method getToStringMethod() throws NoSuchMethodException {
-        return Object.class.getDeclaredMethod("toString");
+    // n.b. cannot use toString because it's special cased
```
do we have tests for special cased `Object` methods `hashCode`, `equals`, and `toString`?

carterkozak replied:
Not yet, wanted to make sure we wanted to move forward first. Will add tests shortly.

schlosna reviewed (commented):
👍 Generally seems fine to me. 

Curious if there's any difference in benchmarks before/after. 

When processing jstacks I generally filter out `"\.(proxy|invoke)"` so don't see these 😄

carterkozak commented:
> Curious if there's any difference in benchmarks before/after.

Locally the simple perf logging benchmark jumped up single-digit nanos, from 14ns to 17ns, but it's difficult to hit the sweet spot on my laptop between the governor spinning up and thermal throttling... I'll try to get better numbers after I've added tests.

> When processing jstacks I generally filter out "\.(proxy|invoke)" so don't see these 😄

If we want to go insane, we can avoid that entirely (spent some time sandboxing with bytebuddy, not suggesting the branch should be merged):
```
Exception in thread "main" java.lang.RuntimeException: Stack Trace
	at com.palantir.DemoMain.run(DemoMain.java:33)
	at com.palantir.tritium.proxy.InstrumentedDemoMain$0.run(Unknown Source)
	at com.palantir.DemoMain.main(DemoMain.java:28)
```
https://github.com/palantir/tritium/compare/ckozak/flatten_tritium_proxy...ckozak/byte_buddy?expand=1

carterkozak committed:
Special method test coverage

carterkozak committed:
Add a simple benchmark with a single handler

before:
```
Benchmark                                         Mode  Cnt   Score   Error  Units
ProxyBenchmark.instrumentedWithEnabledNopHandler  avgt    5  20.608 ± 0.335  ns/op
```

after:
```
Benchmark                                         Mode  Cnt   Score   Error  Units
ProxyBenchmark.instrumentedWithEnabledNopHandler  avgt    5  19.698 ± 0.437  ns/op
```

Benchmarks run on a macbook, trust the results at your own peril.

carterkozak commented:
Benchmarks: I implemented a benchmark with a single handler that passes all arguments to the blackhole to avoid dce. My initial benchmark use an enabled handler with no-op methods, and results were in the 10-11ns range.

before:
```
Benchmark                                         Mode  Cnt   Score   Error  Units
ProxyBenchmark.instrumentedWithEnabledNopHandler  avgt    5  20.608 ± 0.335  ns/op
```

after:
```
Benchmark                                         Mode  Cnt   Score   Error  Units
ProxyBenchmark.instrumentedWithEnabledNopHandler  avgt    5  19.698 ± 0.437  ns/op
```

Benchmarks run on a macbook, trust the results at your own peril.

carterkozak committed:
NullAway

carterkozak commented on tritium-lib/src/test/java/com/palantir/tritium/proxy/InstrumentationTest.java:
```diff
@@ -368,6 +369,51 @@ void testInaccessibleConstructor() throws NoSuchMethodException {
+                .build();
+        assertThat(instrumented).asString().isEqualTo("com.palantir.tritium.test.TestImplementation");
+        verifyNoMoreInteractions(mockHandler);
+    }
```
Tested on develop as well.

carterkozak commented:
I've added relevant testing and benchmarks.

carterkozak commented on tritium-jmh/src/jmh/java/com/palantir/tritium/microbenchmarks/BlackholeInvocationEventHandler.java:
```diff
@@ -0,0 +1,68 @@
+    public void onFailure(@Nullable InvocationContext context, @Nonnull Throwable cause) {
+        consume(context, cause);
+    }
+    private void consume(@Nullable Object obj0, @Nullable Object obj1, @Nullable Object obj2) {
```
oops, missing newline
